In [23]:
import numpy as np
import pandas as pd
import string

import matplotlib.pyplot as plt
%matplotlib inline
plt.style.use('ggplot')
import warnings
warnings.simplefilter('ignore', FutureWarning)

# read table
movies = pd.read_csv('movie_plots.csv')
movies.loc[0:4]

Release Year                             Title Origin/Ethnicity  \
0          1901            Kansas Saloon Smashers         American   
1          1901     Love by the Light of the Moon         American   
2          1901           The Martyred Presidents         American   
3          1901  Terrible Teddy, the Grizzly King         American   
4          1902            Jack and the Beanstalk         American   

                             Director Cast    Genre  \
0                             Unknown  NaN  unknown   
1                             Unknown  NaN  unknown   
2                             Unknown  NaN  unknown   
3                             Unknown  NaN  unknown   
4  George S. Fleming, Edwin S. Porter  NaN  unknown   

                                           Wiki Page  \
0  https://en.wikipedia.org/wiki/Kansas_Saloon_Sm...   
1  https://en.wikipedia.org/wiki/Love_by_the_Ligh...   
2  https://en.wikipedia.org/wiki/The_Martyred_Pre...   
3  https://en.wikipedia.org/wiki/Terrible_Teddy,_...   
4  https://en.wikipedia.org/wiki/Jack_and_the_Bea...   

                                                Plot  
0  A bartender is working at a saloon, serving dr...  
1  The moon, painted with a smiling face hangs ov...  
2  The film, just over a minute long, is composed...  
3  Lasting just 61 seconds and consisting of two ...  
4  The earliest known adaptation of the classic f...

# The Question
The first part of developing a data-driven project is to decide what question you want to answer. The question needs to be specific, and it needs to be something you can develop a step-by-step approach for. With this notebook, I am going to use the `movies` DataFrame to answer the following question:
> Can we predict the genre of a movie based on its plot?

It will take a few steps to answer this question. The over-arching workflow will look something like this:
1. Data preprocessing
2. Group movies by genre and look for recurring words in plots
3. Write a $k$-nearest neighbor classifier
4. Test the classifier and determine its accuracy

# Data Preprocessing
Currently, the `movies` DataFrame contains details of movies (year, title, country of origin, director, cast, genre, Wiki link, and plot). In order to get meaningful results, the data need to be "cleaned;" that is, we need to remove values that we can't work with. The data set, which contains 34,885 entries, has lots of rows that contain `NaN` values ("not a number"), which means that we can't work with them. The `pandas` library allows you to remove the rows that contain `NaN` values:

In [24]:
movies = movies.dropna().reset_index(drop=True)
movies.loc[0:4]

Release Year                            Title Origin/Ethnicity  \
0          1903              Alice in Wonderland         American   
1          1907                     Daniel Boone         American   
2          1907  How Brown Saw the Baseball Game         American   
3          1907                     Laughing Gas         American   
4          1908         The Adventures of Dollie         American   

                                  Director                               Cast  \
0                           Cecil Hepworth                          May Clark   
1  Wallace McCutcheon and Ediwin S. Porter  William Craven, Florence Lawrence   
2                                  Unknown                            Unknown   
3                     Edwin Stanton Porter    Bertha Regustus, Edward Boulden   
4                           D. W. Griffith  Arthur V. Johnson, Linda Arvidson   

          Genre                                          Wiki Page  \
0       unknown  https://en.wikipedia.org/wiki/Alice_in_Wonderl...   
1  biographical  https://en.wikipedia.org/wiki/Daniel_Boone_(19...   
2        comedy  https://en.wikipedia.org/wiki/How_Brown_Saw_th...   
3        comedy  https://en.wikipedia.org/wiki/Laughing_Gas_(fi...   
4         drama  https://en.wikipedia.org/wiki/The_Adventures_o...   

                                                Plot  
0  Alice follows a large white rabbit down a "Rab...  
1  Boone's daughter befriends an Indian maiden as...  
2  Before heading out to a baseball game at a nea...  
3  The plot is that of a black woman going to the...  
4  On a beautiful summer day a father and mother ...

In the above cell, we also reset the row indices to make slicing the table easier. We now have 33,464 rows, so we lost about 1,400 entries. The next thing to look at is what kinds of values we have in the DataFrame. Since we're focusing on the genre and plot, we can remove columns with irrelevant information:

In [25]:
movies = movies[['Genre', 'Plot']]
movies.loc[0:4]

Genre                                               Plot
0       unknown  Alice follows a large white rabbit down a "Rab...
1  biographical  Boone's daughter befriends an Indian maiden as...
2        comedy  Before heading out to a baseball game at a nea...
3        comedy  The plot is that of a black woman going to the...
4         drama  On a beautiful summer day a father and mother ...

### Data Exploration
Now we need to see what kinds of data we have in the set. We begin by showing what the values in the `Genre` column of the DataFrame are:

In [26]:
movies.groupby('Genre').agg(len)

Plot
Genre                                                  
                                                     18
 usa                                                  1
 usa, can                                             1
-                                                     3
16 mm film                                            1
[140]                                                 1
[144]                                                 1
action                                             1087
action                                               21
action & comedy                                       1
action & love                                         2
action & romance                                      3
action (film genre)                                   1
action - comedy                                       1
action - romance                                      2
action - superhero                                    1
action - thriller                                     1
action - war                                          1
action / adventure                                    4
action / adventure / comedy                           2
action / adventure / comedy / crime / drama           1
action / adventure / kung fu                          1
action / adventure / martial arts                     3
action / adventure / suspense                         1
action / adventure / thriller                         1
action / animation / adventure                        1
action / biography / history / war / costume          1
action / comedy                                      14
action / comedy / crime                               2
action / comedy / drama                               2
...                                                 ...
world war ii action                                   1
world war ii comedy                                   4
world war ii documentary                              1
world war ii drama                                    1
world war ii drama documentary                        1
world war ii/adventure                                2
world war ii/comedy                                   1
world war ii/drama                                    4
world war ii/pow                                      4
world war ii/spy                                      1
world war ii/thriller                                 1
wuxia                                                 9
wuxia / romance                                       1
wuxia, comedy                                         1
wuxia/comedy                                          1
wuxia/historical/romance                              1
ww1 propaganda                                        1
wwii drama                                            1
yakuza                                                4
yakuza film                                           1
yash dasgupta, mimi chakraborty                       1
yash raj films                                        2
yeh saali zindagi                                     1
yogesh dattatraya gosavi's directorial debut. [9]     1
youth                                                 1
youth drama                                           1
youth, romance                                        1
yuen lai-kei                                          1
zombie comedy                                         2
zombie horror                                         2

[2210 rows x 1 columns]

It looks like there's lots of different values in the column (2,210 to be exact). But for this question, we're really only interested in the basic genres of movies: action, adventure, comedy, drama, fantasy, historical, horror, romance, science fiction, and thriller.

To this end, we will define a function that will read through the `Genre` entry in each row and categorize them as one of the above. This process has a few steps:
1. Define a function that will determine if any of the above words are present in the entry
2. Filter `movies` for such rows
3. Change the entry for `Genre` to one of the above for each row left if entry is not one of the above

In [27]:
# Step 1:
def acceptable_genre(entry):
    acceptable_genres = ['action', 'adventure', 'comedy', 'drama', 'fantasy', 'historical', 'horror', 
                        'romance', 'science fiction', 'thriller']
    for genre in acceptable_genres:
        if genre in entry['Genre']:
            return True
    return False

# Step 2:
filtered_movies = movies[movies.apply(acceptable_genre, axis=1)].reset_index(drop=True)
filtered_movies.loc[0:4]

Genre                                               Plot
0  comedy  Before heading out to a baseball game at a nea...
1  comedy  The plot is that of a black woman going to the...
2   drama  On a beautiful summer day a father and mother ...
3   drama  A thug accosts a girl as she leaves her workpl...
4  comedy  A young couple decides to elope after being ca...

In [28]:
# Step 3:
def superfluous_text(entry):
    """
    Determines if the string entry is in acceptable_genres
    """
    acceptable_genres = ['action', 'adventure', 'comedy', 'drama', 'fantasy', 'historical', 'horror', 
                        'romance', 'science fiction', 'thriller']
    for genre in acceptable_genres:
        if genre == entry:
            return False
    return True

def determine_genre(entry):
    """
    Takes a string entry and returns the word closest to the end that is in acceptable_genres
    """
    acceptable_genres = ['action', 'adventure', 'comedy', 'drama', 'fantasy', 'historical', 'horror', 
                        'romance', 'science fiction', 'thriller']

    if not superfluous_text(entry):
        return entry
    for c in entry:
        if c not in string.ascii_lowercase:
            genre = entry.replace(c, '')
    try:
        genre = genre.split(' ')
    except UnboundLocalError:
        genre = entry.split(' ')

    i = -1
    new_genre = ''
    while new_genre not in acceptable_genres:
        try:
            genre = genre[i]
        except IndexError:
            new_genre = ''
            return new_genre
        if genre not in acceptable_genres:
            i -= 1

    return new_genre

def change_genres(df):
    new_genres = np.array([])
    for entry in df['Genre']:
        new_genre = determine_genre(entry)
        new_genres = np.append(new_genres, new_genre)
    
    new_df = df.drop(columns=['Genre']).assign(Genre=new_genres)
    
    return new_df

filtered_genres = change_genres(filtered_movies)
filtered_genres.loc[0:4]

Plot   Genre
0  Before heading out to a baseball game at a nea...  comedy
1  The plot is that of a black woman going to the...  comedy
2  On a beautiful summer day a father and mother ...   drama
3  A thug accosts a girl as she leaves her workpl...   drama
4  A young couple decides to elope after being ca...  comedy

### Cleaning Plot Strings


In [47]:
def clean_string(s):
    for c in s:
        if c not in string.ascii_letters + ' ':
            s = s.replace(c, '')
        elif c in string.ascii_uppercase:
            i = string.ascii_uppercase.index(c)
            s = s.replace(c, string.ascii_lowercase[i])
    return s

def clean_plots(df):
    plots = list(df['Plot'])
    
    cleaned_plots = []
    for plot in plots:
        cleaned_plot = clean_string(plot)
        cleaned_plots += [cleaned_plot]
        
    new_df = df.drop(['Plot'], axis=1).assign(Plot=cleaned_plots)
    return new_df

movies_with_cleaned_plots = clean_plots(filtered_genres)
movies_with_cleaned_plots.loc[0:4]

Genre                                               Plot
0  comedy  before heading out to a baseball game at a nea...
1  comedy  the plot is that of a black woman going to the...
2   drama  on a beautiful summer day a father and mother ...
3   drama  a thug accosts a girl as she leaves her workpl...
4  comedy  a young couple decides to elope after being ca...